File: deploy/ios_demo/PaddleOCRDemo/Engine/QuadTextCrop.swift
// Copyright (c) 2026 PaddlePaddle Authors. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import CoreGraphics
import Foundation

// MARK: - BGR ↔ CGImage

/// Builds a ``CGImage`` from row-major BGR uint8 (8-bit three-channel).
enum BGRImageCodec {
    static func makeCGImageFromBGR(pixels: [UInt8], width: Int, height: Int) throws -> CGImage {
        let bytesPerPixel = 4
        let bytesPerRow = width * bytesPerPixel

        let pixelCount = width * height
        guard pixels.count == pixelCount * 3 else {
            throw QuadTextCropError.imageCreationFailed
        }

        var rgbaData = [UInt8](repeating: 255, count: pixelCount * bytesPerPixel)
        for i in 0..<pixelCount {
            let b = pixels[i * 3]
            let g = pixels[i * 3 + 1]
            let r = pixels[i * 3 + 2]
            rgbaData[i * 4] = r
            rgbaData[i * 4 + 1] = g
            rgbaData[i * 4 + 2] = b
        }

        guard let colorSpace = CGColorSpace(name: CGColorSpace.sRGB),
              let context = CGContext(
                  data: &rgbaData,
                  width: width,
                  height: height,
                  bitsPerComponent: 8,
                  bytesPerRow: bytesPerRow,
                  space: colorSpace,
                  bitmapInfo: CGImageAlphaInfo.noneSkipLast.rawValue
              ),
              let cgImage = context.makeImage() else {
            throw QuadTextCropError.imageCreationFailed
        }

        return cgImage
    }
}

// MARK: - Errors

enum QuadTextCropError: LocalizedError {
    case invalidPolygon(String)
    case imageCreationFailed
    case openCVFailed

    var errorDescription: String? {
        switch self {
        case .invalidPolygon(let detail):
            return "Invalid polygon for quad text crop: \(detail)"
        case .imageCreationFailed:
            return "Failed to create output CGImage"
        case .openCVFailed:
            return "OpenCV quad text-line crop failed"
        }
    }
}

// MARK: - QuadTextCrop

/// Minimum-area quadrilateral, corner ordering, perspective warp, then optional 90° CCW rotation when height/width ≥ 1.5.
struct QuadTextCrop {

    /// Crop a text region from the source image using the detector quadrilateral.
    static func crop(_ image: CGImage, polygon: [[Int32]]) throws -> CGImage {
        guard polygon.count == 4, polygon.allSatisfy({ $0.count == 2 }) else {
            throw QuadTextCropError.invalidPolygon(
                "Expected 4 points with 2 coordinates each, got \(polygon.count) points"
            )
        }

        let srcPixels = try extractBGRPixels(from: image)
        let srcW = image.width
        let srcH = image.height

        let polyValues: [NSValue] = polygon.map {
            NSValue(cgPoint: CGPoint(x: CGFloat($0[0]), y: CGFloat($0[1])))
        }
        guard let out = PDBOpenCVImageBridge.quadTextLineCropBGR(
            Data(srcPixels),
            srcWidth: srcW,
            srcHeight: srcH,
            quad: polyValues
        ) else {
            throw QuadTextCropError.openCVFailed
        }

        let bytes = [UInt8](out.rgbData)
        return try BGRImageCodec.makeCGImageFromBGR(pixels: bytes, width: Int(out.width), height: Int(out.height))
    }
}

// MARK: - Pixel Extraction & Image Creation

private extension QuadTextCrop {

    /// Row-major HWC **BGR**
    static func extractBGRPixels(from image: CGImage) throws -> [UInt8] {
        let width = image.width
        let height = image.height
        let bytesPerPixel = 4
        let bytesPerRow = width * bytesPerPixel
        var rgbaData = [UInt8](repeating: 0, count: height * bytesPerRow)

        guard let colorSpace = CGColorSpace(name: CGColorSpace.sRGB),
              let context = CGContext(
                  data: &rgbaData,
                  width: width,
                  height: height,
                  bitsPerComponent: 8,
                  bytesPerRow: bytesPerRow,
                  space: colorSpace,
                  bitmapInfo: CGImageAlphaInfo.noneSkipLast.rawValue
              ) else {
            throw QuadTextCropError.imageCreationFailed
        }

        context.draw(image, in: CGRect(x: 0, y: 0, width: width, height: height))

        let pixelCount = width * height
        var bgrData = [UInt8](repeating: 0, count: pixelCount * 3)
        for i in 0..<pixelCount {
            let r = rgbaData[i * 4]
            let g = rgbaData[i * 4 + 1]
            let b = rgbaData[i * 4 + 2]
            bgrData[i * 3] = b
            bgrData[i * 3 + 1] = g
            bgrData[i * 3 + 2] = r
        }

        return bgrData
    }
}

File: deploy/ppocr-android/ppocr-sdk/src/main/java/com/paddle/ocr/postprocess/QuadTextCrop.kt
// Copyright (c) 2026 PaddlePaddle Authors. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package com.paddle.ocr.postprocess

import com.paddle.ocr.model.OCRBox
import org.opencv.core.Core
import org.opencv.core.CvType
import org.opencv.core.Mat
import org.opencv.core.MatOfPoint2f
import org.opencv.core.Point
import org.opencv.core.Size
import org.opencv.imgproc.Imgproc
import kotlin.math.hypot
import kotlin.math.max

object QuadTextCrop {
    private const val VERTICAL_CROP_RATIO = 1.5

    fun crop(src: Mat, box: OCRBox): Mat {
        // Align with PaddleX CropByPolys.get_minarea_rect_crop: recompute minAreaRect
        // from the detected quad before perspective transform.
        val rectPoints = box.points.map { Point(it.x.toDouble(), it.y.toDouble()) }
        val rectInput = MatOfPoint2f()
        rectInput.fromList(rectPoints)
        val boundingBox = Imgproc.minAreaRect(rectInput)
        rectInput.release()

        val boxPoints = Array(4) { Point() }
        boundingBox.points(boxPoints)
        val ordered = QuadGeometry.orderMinAreaRectPoints(boxPoints)

        val widthTop = hypot(ordered[0].x - ordered[1].x, ordered[0].y - ordered[1].y)
        val widthBottom = hypot(ordered[2].x - ordered[3].x, ordered[2].y - ordered[3].y)
        val heightLeft = hypot(ordered[0].x - ordered[3].x, ordered[0].y - ordered[3].y)
        val heightRight = hypot(ordered[1].x - ordered[2].x, ordered[1].y - ordered[2].y)

        val dstW = max(widthTop, widthBottom).toInt().coerceAtLeast(1)
        val dstH = max(heightLeft, heightRight).toInt().coerceAtLeast(1)

        val srcPts = MatOfPoint2f()
        srcPts.fromList(ordered)
        val dstPts = MatOfPoint2f()
        dstPts.fromList(
            listOf(
                Point(0.0, 0.0),
                Point(dstW.toDouble(), 0.0),
                Point(dstW.toDouble(), dstH.toDouble()),
                Point(0.0, dstH.toDouble()),
            )
        )
        val m = Imgproc.getPerspectiveTransform(srcPts, dstPts)
        srcPts.release()
        dstPts.release()

        val dst = Mat(dstH, dstW, CvType.CV_8UC3)
        Imgproc.warpPerspective(
            src,
            dst,
            m,
            Size(dstW.toDouble(), dstH.toDouble()),
            Imgproc.INTER_CUBIC,
            Core.BORDER_REPLICATE,
        )
        m.release()

        if (dst.rows().toDouble() / dst.cols() >= VERTICAL_CROP_RATIO) {
            val rotated = Mat()
            Core.rotate(dst, rotated, Core.ROTATE_90_COUNTERCLOCKWISE)
            dst.release()
            return rotated
        }
        return dst
    }
}
